### PR TITLE
Change curtain logic, change final thermostat call

### LINKED ISSUE #39: Remove set_value / get_payload_parameter_name
After the last call in VeraThermostat is refactored

### LINKED ISSUE #44: Cover position not updated correctly
Again not sure if this is a pyvera or Hass issue.

But in Hass, if I open my curtain for 50% and then call (full) open (or close) the interface still shows position 50% afterwards. And a side effect is that the controls (arrow up and down) are both enabled, but one should be greyed out.

### PR DESCRIPTION
**Pull request by pavoni**
Changes curtain open / close to set via level - hopefully closes #44 

Change thermostat set_point call to call `set_service_value` closes #39 

Removes `set_value` and `get_payload_parameter_name` . This was the final instance of the 'try and guess the parameter names' logic so it's now gone!

@arjenfvellinga Could you test whether #44 is fixed? 
Details on how to run development HA are here (https://home-assistant.io/developers/development_environment/) 
How to use custom library version are at the bottom of this page (https://home-assistant.io/developers/component_deps_and_reqs/)

You will need this branch of HA https://github.com/home-assistant/home-assistant/tree/update_vera_cover_refresh_logic

And the the new version of pyvera (for the custom library install) is `0.2.25 `

@jnewland @robjohnson189 @philhawthorne Would one of you mind checking that I didn't break the ability to set a thermostat temperature?

### CODE REVIEW

**Comment by arjenfvellinga** [comment]
I will have a look, probably this Friday

**Comment by arjenfvellinga** [comment]
I've tested your branch and it looks good at a user perspective. A full open / close updates the control accordingly.

Here is the logging of closing the cover (gordijnen in Dutch) to 57% and afterwards using the open command. It sets the position back to 100%.

```
17-04-01 14:08:38 DEBUG (MainThread) [homeassistant.components.websocket_api] WS 139926300205520: Received {'type': 'call_service', 'id': 9, 'service': 'set_cover_position', 'domain': 'cover', 'service_data': {'entity_id': 'cover.gordijnen_21', 'position': 57}}
17-04-01 14:08:39 INFO (Vera Poll Thread) [pyvera.subscribe] Poll returned
17-04-01 14:08:39 DEBUG (Thread-8) [pyvera] set_service_value: result of vera_request with payload {'newLoadlevelTarget': 57, 'action': 'SetLoadLevelTarget', 'id': 'lu_action', 'serviceId': 'urn:upnp-org:serviceId:Dimming1'}: { "u:SetLoadLevelTargetResponse": { "JobID": "1744" } }
17-04-01 14:08:39 DEBUG (MainThread) [pyvera] get_level: False : 57
17-04-01 14:08:39 DEBUG (MainThread) [pyvera] get_level: False : 57
17-04-01 14:08:39 DEBUG (MainThread) [pyvera] get_level: False : 57
17-04-01 14:08:39 DEBUG (MainThread) [pyvera] get_level: False : 57
17-04-01 14:08:39 DEBUG (MainThread) [pyvera] get_level: False : 57
17-04-01 14:08:39 DEBUG (MainThread) [homeassistant.components.websocket_api] WS 139926300205520: Sending {'event': {'event_type': 'state_changed', 'data': {'new_state': <state cover.gordijnen_21=open; supported_features=15, current_position=57, Vera Device Id=21, friendly_name=Gordijnen @ 2017-04-01T14:07:03.469525+02:00>, 'old_state': <state cover.gordijnen_21=open; supported_features=15, current_position=100, Vera Device Id=21, friendly_name=Gordijnen @ 2017-04-01T14:07:03.469525+02:00>, 'entity_id': 'cover.gordijnen_21'}, 'origin': 'LOCAL', 'time_fired': datetime.datetime(2017, 4, 1, 12, 8, 39, 26849, tzinfo=<UTC>)}, 'type': 'event', 'id': 2}
17-04-01 14:08:39 DEBUG (MainThread) [homeassistant.components.websocket_api] WS 139926300253488: Sending {'event': {'event_type': 'state_changed', 'data': {'new_state': <state cover.gordijnen_21=open; supported_features=15, current_position=57, Vera Device Id=21, friendly_name=Gordijnen @ 2017-04-01T14:07:03.469525+02:00>, 'old_state': <state cover.gordijnen_21=open; supported_features=15, current_position=100, Vera Device Id=21, friendly_name=Gordijnen @ 2017-04-01T14:07:03.469525+02:00>, 'entity_id': 'cover.gordijnen_21'}, 'origin': 'LOCAL', 'time_fired': datetime.datetime(2017, 4, 1, 12, 8, 39, 26849, tzinfo=<UTC>)}, 'type': 'event', 'id': 2}
17-04-01 14:08:39 DEBUG (MainThread) [homeassistant.components.websocket_api] WS 139926300205520: Sending {'result': None, 'type': 'result', 'id': 9, 'success': True}
17-04-01 14:08:40 INFO (Vera Poll Thread) [pyvera.subscribe] Poll returned
17-04-01 14:08:40 DEBUG (MainThread) [pyvera] get_level: False : 57
17-04-01 14:08:40 DEBUG (MainThread) [pyvera] get_level: False : 57
17-04-01 14:08:40 DEBUG (MainThread) [pyvera] get_level: False : 57
17-04-01 14:08:40 DEBUG (MainThread) [pyvera] get_level: False : 57
17-04-01 14:08:40 DEBUG (MainThread) [pyvera] get_level: False : 57
17-04-01 14:08:51 DEBUG (MainThread) [homeassistant.components.websocket_api] WS 139926300205520: Received {'type': 'call_service', 'id': 10, 'service': 'open_cover', 'domain': 'cover', 'service_data': {'entity_id': 'cover.gordijnen_21'}}
17-04-01 14:08:51 INFO (Vera Poll Thread) [pyvera.subscribe] Poll returned
17-04-01 14:08:51 DEBUG (Thread-11) [pyvera] set_service_value: result of vera_request with payload {'newLoadlevelTarget': 100, 'action': 'SetLoadLevelTarget', 'id': 'lu_action', 'serviceId': 'urn:upnp-org:serviceId:Dimming1'}: { "u:SetLoadLevelTargetResponse": { "JobID": "1745" } }
17-04-01 14:08:51 DEBUG (MainThread) [pyvera] get_level: False : 100
17-04-01 14:08:51 DEBUG (MainThread) [pyvera] get_level: False : 100
17-04-01 14:08:51 DEBUG (MainThread) [pyvera] get_level: False : 100
17-04-01 14:08:51 DEBUG (MainThread) [pyvera] get_level: False : 100
17-04-01 14:08:51 DEBUG (MainThread) [pyvera] get_level: False : 100
17-04-01 14:08:51 DEBUG (MainThread) [homeassistant.components.websocket_api] WS 139926300205520: Sending {'event': {'event_type': 'state_changed', 'data': {'new_state': <state cover.gordijnen_21=open; supported_features=15, current_position=100, Vera Device Id=21, friendly_name=Gordijnen @ 2017-04-01T14:07:03.469525+02:00>, 'old_state': <state cover.gordijnen_21=open; supported_features=15, current_position=57, Vera Device Id=21, friendly_name=Gordijnen @ 2017-04-01T14:07:03.469525+02:00>, 'entity_id': 'cover.gordijnen_21'}, 'origin': 'LOCAL', 'time_fired': datetime.datetime(2017, 4, 1, 12, 8, 51, 633653, tzinfo=<UTC>)}, 'type': 'event', 'id': 2}
17-04-01 14:08:51 DEBUG (MainThread) [homeassistant.components.websocket_api] WS 139926300253488: Sending {'event': {'event_type': 'state_changed', 'data': {'new_state': <state cover.gordijnen_21=open; supported_features=15, current_position=100, Vera Device Id=21, friendly_name=Gordijnen @ 2017-04-01T14:07:03.469525+02:00>, 'old_state': <state cover.gordijnen_21=open; supported_features=15, current_position=57, Vera Device Id=21, friendly_name=Gordijnen @ 2017-04-01T14:07:03.469525+02:00>, 'entity_id': 'cover.gordijnen_21'}, 'origin': 'LOCAL', 'time_fired': datetime.datetime(2017, 4, 1, 12, 8, 51, 633653, tzinfo=<UTC>)}, 'type': 'event', 'id': 2}
17-04-01 14:08:51 DEBUG (MainThread) [homeassistant.components.websocket_api] WS 139926300205520: Sending {'result': None, 'type': 'result', 'id': 10, 'success': True}
17-04-01 14:08:52 INFO (Vera Poll Thread) [pyvera.subscribe] Poll returned
17-04-01 14:08:52 DEBUG (MainThread) [pyvera] get_level: False : 100
17-04-01 14:08:52 DEBUG (MainThread) [pyvera] get_level: False : 100
17-04-01 14:08:52 DEBUG (MainThread) [pyvera] get_level: False : 100
17-04-01 14:08:52 DEBUG (MainThread) [pyvera] get_level: False : 100
17-04-01 14:08:52 DEBUG (MainThread) [pyvera] get_level: False : 100
17-04-01 14:09:08 INFO (Vera Poll Thread) [pyvera.subscribe] Poll returned
```

I do notice that get_level is requested 5 times in a row, but maybe that is a HA thing.

And this is a part of the startup logging:

```
17-04-01 14:07:03 INFO (Thread-6) [pyvera.subscribe] Subscribing to events for Buitenlamp
17-04-01 14:07:03 INFO (Thread-10) [pyvera.subscribe] Subscribing to events for Gordijnen
17-04-01 14:07:03 INFO (Thread-6) [pyvera.subscribe] Subscribing to events for Lamp bij de bank
17-04-01 14:07:03 DEBUG (MainThread) [pyvera] get_level: False : 100
17-04-01 14:07:03 INFO (Thread-6) [pyvera.subscribe] Subscribing to events for Spotjes boven de tafel
17-04-01 14:07:03 DEBUG (MainThread) [pyvera] get_level: False : 100
17-04-01 14:07:03 INFO (Thread-6) [pyvera.subscribe] Subscribing to events for Plafondlamp
```
Without looking into it, I should expect get_level also here for covers?

File: pyvera/__init__.py [hunk 1]
```diff
@@ -294,10 +294,15 @@ def thermostat_fan_service(self):
...
         return 'urn:upnp-org:serviceId:HVAC_FanOperatingMode1'
 
-    # pylint: disable=R0201
-    def get_payload_parameter_name(self, name):
-        """the http payload for setting a variable"""
-        return 'new' + name + 'Value'
+    @property
+    def thermostat_cool_setpoint(self):
+        """Vera service string Temperature Setpoint Heat."""
+        return 'urn:upnp-org:serviceId:TemperatureSetpoint1_Heat'
```
Comment (jnewland) [inline pyvera/__init__.py]: Looks like the names of these properties are inverted?

File: pyvera/__init__.py [hunk 2]
```diff
@@ -731,12 +716,22 @@ def is_locked(self, refresh=False):
...
     def set_temperature(self, temp):
         """Set current goal temperature / setpoint"""
-        self.set_value('CurrentSetpoint', temp)
+
+        # Set both the heat and cool setpointa
+        self.set_service_value(
+            self.thermostat_cool_setpoint,
+            'CurrentSetpoint',
+            'NewCurrentSetpoint',
+            mode)
```
Comment (jnewland) [inline pyvera/__init__.py]: `s/mode/temp`, same below.

File: pyvera/__init__.py [hunk 3]
```diff
@@ -731,12 +716,22 @@ def is_locked(self, refresh=False):
...
+
+        self.set_service_value(
+            self.thermostat_heat_setpoint,
+            'CurrentSetpoint',
+            'NewCurrentSetpoint',
+            mode)
```
Comment (jnewland) [inline pyvera/__init__.py]: I'm not sure this will always be the correct action to take. I've been working through some confusion with the Thermostat API with Vera support, and I'm still a bit confused TBH. Here are some quotes from our support thread:

> Auto is not a mode which you should set up manually, would be better if you could set your Heat and Cool setpoints and then leave Auto to work on it’s own making the average and running on that value.

> Also, in order to work properly, for some thermostats the different between Heat and Cool must be greater than 3 degrees, please try to enlarge the gap between Heat and Cool and then give Auto a try.

In a non-auto mode, setting `CurrentSetpoint` seems like the appropriate action to take. When a thermostat is in Auto mode, it almost sounds like they want the following actions to take place to make a change:

* Set the operating mode to some non-auto value
* Update the heat and cool values, making sure they're least 3 degrees apart
* Set the operating mode back to the existing mode

So, in short, this isn't right, but it's clear to me that I need to do some more testing before I'm able to tell you what to change it to.

Comment (pavoni) [inline pyvera/__init__.py]: Thanks very much for taking a look - and fixing the bugs. Hopefully now fixed.

If the bugs are fixed than I'm pretty sure this will do what the old code did (set everything with a `setpoint` to the new value). 

I'm sure this isn't a good solution for a heat/cool heating system - but it probably `sort of` works.

This at least uncovers the issue (which was hidden by the old code). As long as it works as well as it did before - I'd prefer to release.

If you have the time inclination - then it would be great to have another PR to make this correct. This will doubtless need changes to how we interact with HA too.

Does that make sense to you? 

Also be very grateful if you could check the code now works!

**Comment by pavoni** [comment]
@arjenfvellinga Thanks for testing. The multiple calls to `get_level` aren't great code - but they're harmless. The `false` parameter means the state is read from the `pyvera` cache - and HA filters out unchanged state update. It's probably some of my HA code - so will take a look.

I'm not sure the missing `get_level` is an issue. It may just be a subscription update following the subscription for the other devices. Does your blind get the correct state when you first start HA?

**Comment by jnewland** [comment]
> If the bugs are fixed than I'm pretty sure this will do what the old code did (set everything with a setpoint to the new value).

[Here's a gist](https://gist.github.com/jnewland/dec09b2bb548095fcd31c8dfe3c6362e) that shows the results of some actions intended to simulate the API calls this PR will make and their effect on the status of the device both via observation and via Vera API calls.

In summary, calling both the heat and cool setpoints has different results in all three modes:

| Action / Mode / Result | *AutoChangeOver* | *CoolOn* | *HeatOn* |
| --- | --- | --- | --- | 
| **Set Cool Setpoint** | No change | Set temp to value submitted | Turn on cool, set to value submitted |
| **Set Heat Setpoint** | No change | Turn on heat, set to value submitted | Set temp to value submitted |

Specifically, *setting `org:serviceId:TemperatureSetpoint1_Heat` changes the mode to `HeatOn` and setting `org:serviceId:TemperatureSetpoint1_Cool` changes the mode to `CoolOn`*. Changing them both back to back flips the unit between two modes quickly and creates unexpected results - someone might try to adjust the AC from 74 to 73 with this new approach and end up with Heat set to 73. That will break this library for anyone using [this thermostat](https://www.amazon.com/GoControl-Thermostat-Z-Wave-Battery-Powered-Amazon/dp/B00ZIRV40K) at the lest.

I'll have some more time to test this next weekend, but I'm running out of time now. Would you be willing to switch back to updating `urn:upnp-org:serviceId:TemperatureSetpoint1` for the time being since this change appears to have unintended side-effects?

**Comment by pavoni** [comment]
Of course - certainly don't want to break things, particularly in code I can't test myself!

Do you think we should just call `urn:upnp-org:serviceId:TemperatureSetpoint1 ` for now?

This is a one log of what the old code did, so I guess it did all three!

````
17-02-27 20:46:04 DEBUG (Thread-12) [pyvera] Set variable: CurrentSetpoint using service_id: urn:upnp-org:serviceId:TemperatureSetpoint1
17-02-27 20:46:04 DEBUG (Thread-12) [pyvera] Result of vera_request with payload {'id': 'lu_action', 'serviceId': 'urn:upnp-org:serviceId:TemperatureSetpoint1', 'action': 'SetCurrentSetpoint', 'NewCurrentSetpoint': 34.0}: { "u:SetCurrentSetpointResponse": { "JobID": "17875" } }
17-02-27 20:46:04 DEBUG (Thread-12) [pyvera] Set variable: CurrentSetpoint using service_id: urn:upnp-org:serviceId:TemperatureSetpoint1_Heat
17-02-27 20:46:04 DEBUG (Thread-12) [pyvera] Result of vera_request with payload {'id': 'lu_action', 'serviceId': 'urn:upnp-org:serviceId:TemperatureSetpoint1_Heat', 'action': 'SetCurrentSetpoint', 'NewCurrentSetpoint': 34.0}: { "u:SetCurrentSetpointResponse": { "JobID": "17876" } }
17-02-27 20:46:04 DEBUG (Thread-12) [pyvera] Set variable: CurrentSetpoint using service_id: urn:upnp-org:serviceId:TemperatureSetpoint1_Cool
17-02-27 20:46:04 DEBUG (Thread-12) [pyvera] Result of vera_request with payload {'id': 'lu_action', 'serviceId': 'urn:upnp-org:serviceId:TemperatureSetpoint1_Cool', 'action': 'SetCurrentSetpoint', 'NewCurrentSetpoint': 34.0}: { "u:SetCurrentSetpointResponse": { "JobID": "17877" } }

````

**Comment by pavoni** [comment]
And thanks for taking the time @jnewland - very grateful.

**Comment by jnewland** [comment]
> This is a one log of what the old code did, so I guess it did all three!

The logs of the existing code explain some odd behavior I've seen but not yet reported. 😬 

> Do you think we should just call urn:upnp-org:serviceId:TemperatureSetpoint1 for now?

Yeah I think so. It won't do anything in Auto mode, but will do the right thing in Heat and Cool modes, which is an improvement over how things are working at the moment.

**Comment by jnewland** [comment]
> And thanks for taking the time @jnewland - very grateful.

🙇 and thank you for spending the time to write software for a device you don't have! I'm happy to follow up on this change with more testing to improve the situation further. We should definitely be able to paper over Vera's weird UX (including their advice to just not touch the thermostat in Auto mode?) with some time and testing to figure out what APIs are the right ones to call in which mode.

**Comment by pavoni** [comment]
@jnewland Just to check - do you think we should merge / release the current code (with the single `setpoint` set) - or wait until we can do some more work?

**Comment by jnewland** [comment]
Hey sorry I wasn't clear! Let's merge this since it's an improvement. I'll do some tests soon and get issues filed for future improvements.
